packages/discovery-provider/plugins/pedalboard/apps/relay/src/middleware/validator.ts

```diff
@@ -50,23 +49,50 @@ export const validator = async (
 
   // Gather user from input data
   // @ts-ignore, partially populate for now
-  let user: Users = {
+  let recoveredSigner: Users | DeveloperApps = {
     wallet: senderAddress || null,
     handle: handle || null
   }
-  try {
-    user = await retrieveUser(
-      contractRegistryKey,
-      contractAddress,
-      encodedABI,
-      senderAddress,
-      handle
-    )
-  } catch (e) {
-    logger.error(
-      { e },
-      'could not gather user from db, continuing with senderAddress and handle'
-    )
+
+  let signerIsApp = false
+  let signerIsUser = false
+  let createOrDeactivate = false
+
+  const user = await retrieveUser(
+    contractRegistryKey,
+    contractAddress,
+    encodedABI,
+    senderAddress,
+    handle
+  )
+  if (user !== undefined) {
+    recoveredSigner = user
+    signerIsUser = true
+  }
+
+  if (signerIsUser) {
+    const isDeactivated = (recoveredSigner as Users).is_deactivated
+    if (isUserDeactivate(isDeactivated, encodedABI)) {
+      logger.info({ requestId: response.locals.ctx.requestId, encodedABI }, "user deactivation")
+      createOrDeactivate = true
+    }
+  }
+
+  if (isUserCreate(encodedABI)) {
+    logger.info({ requestId: response.locals.ctx.requestId, encodedABI }, "user create")
+    createOrDeactivate = true
+  }
+
+  // could not find user and is not create, find app
+  if (!signerIsUser && !createOrDeactivate) {
+    const developerApp = await retrieveDeveloperApp({ encodedABI, contractAddress })
+    if (developerApp === undefined) {
+      logger.error({ encodedABI }, "neither user nor developer app could be found for address")
```

Comment:
refactored this quite a bit, here's basically what happens:
1. gather user, if no user: continue. otherwise signer is user
2. if signer is not a user check if we're creating one, if so mark this as true
3. if signer is a user, check if they are deactivating said user
4. if signer is not a user nor a create it must be an app, retrieve the app
5. if not an app, user, or creation, or deactivation, then give a validation error